File: LeetCode/Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
	# @param inorder, a list of integers
	# @param postorder, a list of integers
	# @return a tree node
	def buildTree(self, inorder, postorder):
		if len(postorder) == 0:
			return None

		head = self.convert(inorder, postorder, 0, len(postorder) - 1, len(postorder) - 1)

		return head

	def convert(self, inorder, postorder, st_idx, end_idx, root_idx):
		if st_idx > end_idx:
			return None 
		tnode = TreeNode(postorder[root_idx])
		mid = st_idx
		for i in xrange(st_idx, end_idx + 1):
			if inorder[i] == tnode.val:
				mid = i
				break

		tnode.left = self.convert(inorder, postorder, st_idx, mid - 1, root_idx - 1 - end_idx + mid)
		tnode.right = self.convert(inorder, postorder, mid + 1, end_idx, root_idx - 1)

		return tnode

if __name__ == '__main__':
	s = Solution()
	# inorder = [1,3,2]
	# postorder = [3,2,1]
	inorder = [4, 2, 5, 1, 3, 6]
	postorder = [4, 5, 2, 6, 3, 1]
	head = s.buildTree(inorder, postorder)
	print head.left.val
	print head.right.val
	print head.left.left.val
	print head.left.right.val
	print head.right.left.val

File: LeetCode/Construct Binary Tree from Inorder and Postorder Traversal.cpp
#include <vector>
#include <stddef.h>
using namespace std;
/**
 * Definition for binary tree
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
private:
    vector<int> io;
    vector<int> po;
public:
    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
        io = inorder;
        po = postorder;
        int inl = io.size(), pol = po.size();
        return buildTree(0, inl-1, 0, pol - 1);
    }

    TreeNode *buildTree(int io_start, int io_end, int po_start, int po_end){
        if (io_start > io_end || po_start > po_end)
            return NULL;
        int idx, rt = po[po_end];
        TreeNode *root = new TreeNode(rt);
        for (int i = io_start; i <= io_end; i++)
            if (io[i] == rt){
                idx = i;
                break;
            }
        root->left = buildTree(io_start, idx-1, po_start, po_start + idx - io_start - 1);
        root->right = buildTree(idx+1, io_end, po_start+idx-io_start, po_end-1);
        return root;
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;
    vector<int> a,b;
    a.push_back(2);
    a.push_back(1);
    b.push_back(2);
    b.push_back(1);
    sol.buildTree(a,b);
    return 0;
}